src/plugins/gs-plugin-eos.c

```diff
@@ -930,3 +939,169 @@ gs_plugin_refresh (GsPlugin *plugin,
 	reload_default_branches (plugin);
 	return TRUE;
 }
+
+static gint
+g_slist_compare_app_ids (GsApp *a,
+			 GsApp *b)
+{
+	return g_strcmp0 (gs_app_get_id (a), gs_app_get_id (b));
+}
+
+static gboolean
+filter_proxied_apps (GsApp *app,
+		     gpointer data)
+{
+	GSList *proxied_apps = (GSList *) data;
+	return g_slist_find_custom (proxied_apps, app,
+				(GCompareFunc) g_slist_compare_app_ids) == NULL;
+}
+
+static GsApp *
+gs_plugin_eos_create_updates_proxy_app (GsPlugin *plugin)
+{
+	const char *id = EOS_PROXY_APP_PREFIX ".EOSUpdatesProxy";
+	const char *unique_id;
+	GsApp *proxy = gs_app_new (id);
+	g_autoptr(AsIcon) icon;
+
+	unique_id = g_strdup_printf ("system/proxy/*/runtime/%s/*/*", id);
+
+	gs_app_set_unique_id (proxy, unique_id);
+	gs_app_set_kind (proxy, AS_APP_KIND_RUNTIME);
+	/* TRANSLATORS: this is the name of the Endless Platform app */
+	gs_app_set_name (proxy, GS_APP_QUALITY_NORMAL,
+			 _("Endless Platform"));
+	/* TRANSLATORS: this is the summary of the Endless Platform app */
+	gs_app_set_summary (proxy, GS_APP_QUALITY_NORMAL,
+			    _("Framework for applications"));
+	gs_app_set_state (proxy, AS_APP_STATE_UPDATABLE_LIVE);
+	gs_app_set_management_plugin (proxy, gs_plugin_get_name (plugin));
+
+	icon = as_icon_new ();
+	as_icon_set_kind (icon, AS_ICON_KIND_STOCK);
+	as_icon_set_name (icon, "system-run-symbolic");
+	gs_app_add_icon (proxy, icon);
+
+	return proxy;
+}
+
+static GsFlatpak *
+gs_plugin_get_gs_flatpak_for_app (GsPlugin *plugin,
+				  GsApp *app)
+{
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+	const char *app_unique_id = gs_app_get_unique_id (app);
+
+	if (g_str_has_prefix (app_unique_id, GS_FLATPAK_SYSTEM_PREFIX))
+		return priv->sys_flatpak;
+
+	return priv->usr_flatpak;
+}
+
+gboolean
+gs_plugin_update_app (GsPlugin *plugin,
+		      GsApp *proxy,
+		      GCancellable *cancellable,
+		      GError **error)
+{
+	GPtrArray *proxied_apps = NULL;
+	guint i;
+	guint num_apps_updated = 0;
+	guint num_apps_to_update = 0;
+
+	/* we only update proxy apps in this plugin */
+	if (!app_is_proxy (proxy))
+		return TRUE;
+
+	proxied_apps = gs_app_get_related (proxy);
+
+	if (proxied_apps->len == 0)
+		return TRUE;
+
+	gs_app_set_state (proxy, AS_APP_STATE_INSTALLING);
+
+	num_apps_to_update = proxied_apps->len;
+	for (i = 0; i < num_apps_to_update; ++i) {
+		GsApp *app = g_ptr_array_index (proxied_apps, i);
+		GsFlatpak *flatpak = gs_plugin_get_gs_flatpak_for_app (plugin,
+								       app);
+		const char *management;
+		gboolean update_result = FALSE;
+
+		g_debug ("Updating '%s' from proxy '%s' ",
+			 gs_app_get_unique_id (app),
+			 gs_app_get_unique_id (proxy));
+
+		/* set the management plugin momentaneously so we can really
+		 * update it; we reset it back later */
+		management = gs_app_get_management_plugin (app);
```

Comment:
You may need to strdup this, or it may be freed when you call `gs_app_set_management_plugin()` right after...